src/Sarif.Multitool/Rules/SarifValidationSkimmerBase.cs

```diff
@@ -57,18 +57,38 @@ protected void LogResult(string jPointer, string formatId, params string[] args)
                 RuleUtilities.BuildResult(DefaultLevel, Context, region, formatId, argsWithPointer));
         }
 
+        protected virtual void Analyze(Attachment attachment, string attachmentPointer)
```

Comment:
>attachment [](start = 50, length = 10)

A bunch of new objects I have to visit to implement the restored rules in SARIF v2. If you haven't seen this code before, here's how it works:

There are a bunch of private `Visit` methods, below, that are responsible for walking down the object tree. Each `Visit` method first calls one of these virtual `Analyze` methods to validate the object at this level, and then drills down into the object model by calling the `Visit` method on its member objects. Individual `Rule` classes override the appropriate `Analyze` method(s) to implement their validation logic.

Reply:
Really we ought to be able to code gen this visitor.

---
In reply to: [210441226](https://github.com/Microsoft/sarif-sdk/pull/1018#discussion_r210441226) [](ancestors = 210441226)